### DISCUSSION TITLE
Funqy dynamic json parsing

### DISCUSSION THREAD

**Discussion by vladaman** [quarkusio/quarkus#29430] [Q&A]
Is there any way to conditionally parse different event types with Funqy?

We use Funqy to process multiple AWS Cognito events such as `CognitoUserPoolPreAuthenticationEvent` or `CognitoUserPoolPostAuthenticationEvent`.

We'd like to have single handler and single running function and branching based on the event type.

We tried following approach but getting JSON as string doesn't work here. Is there any solution to this?

```java
   @Funq("handleString")
    public Uni<? extends CognitoUserPoolEvent> handleString(String cognitoEvent) {
        ObjectMapper objectMapper = new ObjectMapper();
        try {
            CognitoUserPoolEvent baseEvent = objectMapper.readValue(cognitoEvent, CognitoUserPoolEvent.class);
            if (baseEvent.getTriggerSource().equals("PreAuthentication_Authentication")) {
                return preAuthentication(objectMapper.readValue(cognitoEvent, CognitoUserPoolPreAuthenticationEvent.class));
            } else if (baseEvent.getTriggerSource().equals("PostAuthentication_Authentication")) {
                return postAuthentication(objectMapper.readValue(cognitoEvent, CognitoUserPoolPostAuthenticationEvent.class));
            } else if (baseEvent.getTriggerSource().equals("PreSignUp_SignUp")) {
                return preSignup(objectMapper.readValue(cognitoEvent, CognitoUserPoolPreSignUpEvent.class));
            } else {
                return Uni.createFrom().item(baseEvent);
            }
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
```

[a comment by an automated account is not shown]

**Comment by matejvasek** [comment]
Does't `byte[]` work?

**Comment by vladaman** [comment]
Hi @matejvasek thanks! I tried `byte[]` but I get an error

```
START RequestId: 8622cd53-5958-4b45-a3e4-2a2a8cdcf55c Version: $LATEST
Cannot deserialize value of type `[B` from Object value (token `JsonToken.START_OBJECT`)
 at [Source: (ByteArrayInputStream); line: 1, column: 1]: com.fasterxml.jackson.databind.exc.MismatchedInputException
com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `[B` from Object value (token `JsonToken.START_OBJECT`)
 at [Source: (ByteArrayInputStream); line: 1, column: 1]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
```


My code:

```java
    @Funq("handle")
    public Uni<? extends CognitoUserPoolEvent> handleString(byte[] cognitoPayload) {
        ObjectMapper mapper = new ObjectMapper();
        try {
            CognitoUserPoolEvent userPoolEvent = mapper.readValue(cognitoPayload, CognitoUserPoolEvent.class);
```

**Comment by matejvasek** [comment]
@vladaman I will need to check that. I pretty sure `byte[]` should work with HTTP/Knative binding, not sure above aws/google bindings.

**Reply by matejvasek** [reply]
@vladaman what binding do you use? aws lambda?

**Reply by vladaman** [reply]
@matejvasek we use `quarkus-amazon-lambda-http`

**Comment by patriot1burke** [comment]
If you are using Lambda constructs like Cognito, why not just use Lambda?

**Comment by patriot1burke** [comment]
byte[] will not work with funqy-lambda.  We need to fix that.  Not sure why String doesn't work for you though.  You aren't getting json?

**Reply by matejvasek** [reply]
I believe that with `String` you need to pass actual JSON string as body, i.e. `"{}"`.

**Reply by matejvasek** [reply]
I mean including begin/end quotes and escaping.

**Comment by matejvasek** [comment]
@patriot1burke draft https://github.com/matejvasek/quarkus/commit/d43743f1f3ce99f604454fac5e2ed6a2c8b90925

**Reply by matejvasek** [reply]
@vladaman could you try the code above?

**Reply by patriot1burke** [reply]
> @patriot1burke draft [matejvasek@d43743f](https://github.com/matejvasek/quarkus/commit/d43743f1f3ce99f604454fac5e2ed6a2c8b90925)

looks good Matej!

**Reply by vladaman** [reply]
@matejvasek I would love to! Unfortunately I don't know how to use your github repository in our project to test out. Is there any guide? Thank you.

**Comment by matejvasek** [comment]
@vladaman btw do you use `native-image` build? I suspect that with `Uni<? extends CognitoUserPoolEvent>` Quarkus will not know what classes register for reflection.

**Reply by vladaman** [reply]
I don't have experience with native images & lambda. So we deploy target/function.zip to lambda as bunch of jar files. Cold start is 7s for some reason. So we prefer to have one function to handle multiple scenarios with Funqy to minimize number of cold starts needed.

We package functions with `mvn clean package`

`Duration: 8519.28 ms Billed Duration: 8520 ms Memory Size: 1024 MB Max Memory Used: 209 MB Init Duration: 2328.38 ms`

List of included artifacts:
```xml
      <artifactId>url-connection-client</artifactId>
      <artifactId>quarkus-funqy-amazon-lambda</artifactId>
      <artifactId>quarkus-junit5</artifactId>
      <artifactId>quarkus-test-amazon-lambda</artifactId>
      <artifactId>aws-java-sdk-cognitoidp</artifactId>
      <artifactId>jnanoid</artifactId>
      <artifactId>eventbridge</artifactId>
      <artifactId>quarkus-maven-plugin</artifactId>
      <artifactId>maven-compiler-plugin</artifactId>
      <artifactId>maven-surefire-plugin</artifactId>
      <artifactId>maven-failsafe-plugin</artifactId>
```

Do you think moving to native image would help?

**Reply by vladaman** [reply]
So I tried native compilation which results in many errors. I guess it requires major code refactoring from our side to make it work. But I guess that would be for another discussion.
